tests/test_template_instantiator.py

```diff
@@ -0,0 +1,606 @@
+"""
+GTSAM Copyright 2010-2020, Georgia Tech Research Corporation,
+Atlanta, Georgia 30332-0415
+All Rights Reserved
+
+See LICENSE for the license information
+
+Tests for template_instantiator.
+
+Author: Varun Agrawal
+"""
+
+# pylint: disable=import-error,wrong-import-position
+
+import os
+import sys
+import unittest
+
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+from gtwrap.interface_parser import (Argument, ArgumentList, Class,
+                                     Constructor, ForwardDeclaration,
+                                     GlobalFunction, Include, Method,
+                                     Namespace, ReturnType, StaticMethod,
+                                     Typename)
+from gtwrap.template_instantiator import (
+    InstantiatedClass, InstantiatedConstructor, InstantiatedDeclaration,
+    InstantiatedGlobalFunction, InstantiatedMethod, InstantiatedStaticMethod,
+    InstantiationHelper, instantiate_args_list, instantiate_name,
+    instantiate_namespace, instantiate_return_type, instantiate_type,
+    is_scoped_template)
+
+
+class TestInstantiationHelper(unittest.TestCase):
+    """Tests for the InstantiationHelper class."""
+    def test_constructor(self):
+        """Test constructor."""
+        helper = InstantiationHelper(InstantiatedClass)
+        self.assertEqual(helper.instantiation_type, InstantiatedClass)
+        helper = InstantiationHelper(InstantiatedConstructor)
+        self.assertEqual(helper.instantiation_type, InstantiatedConstructor)
+        helper = InstantiationHelper(InstantiatedDeclaration)
+        self.assertEqual(helper.instantiation_type, InstantiatedDeclaration)
+        helper = InstantiationHelper(InstantiatedGlobalFunction)
+        self.assertEqual(helper.instantiation_type, InstantiatedGlobalFunction)
+        helper = InstantiationHelper(InstantiatedMethod)
+        self.assertEqual(helper.instantiation_type, InstantiatedMethod)
+        helper = InstantiationHelper(InstantiatedStaticMethod)
+        self.assertEqual(helper.instantiation_type, InstantiatedStaticMethod)
+
+    def test_instantiate(self):
+        """Test instantiate method."""
+        method = Method.rule.parseString("""
+            template<U={double}>
+            double method(const T x, const U& param);
+            """)[0]
+        cls = Class.rule.parseString("""
+            template<T={string}>
+            class Foo {};
+        """)[0]
+        typenames = ['T', 'U']
+        class_instantiations = [Typename.rule.parseString("string")[0]]
+        method_instantiations = [Typename.rule.parseString("double")[0]]
```

Comment:
I'm kind of confused why these need to be explicitly defined; don't they get parsed by `original.template.instantiations` ?  Or rather, is it true that `class_instantiations` should be the same as `cls.template.instantiations` ?